cpp/src/table/row_operators.cu

```diff
@@ -373,51 +385,391 @@ namespace row {
+  auto d_table = table_device_view::create(preprocessed_input, stream);
+  auto d_column_order =
+    detail::make_device_uvector_async(column_order, stream, rmm::mr::get_current_device_resource());
   auto d_null_precedence = detail::make_device_uvector_async(
-    new_null_precedence, stream, rmm::mr::get_current_device_resource());
+    null_precedence, stream, rmm::mr::get_current_device_resource());
   auto d_depths = detail::make_device_uvector_async(
     verticalized_col_depths, stream, rmm::mr::get_current_device_resource());
 
-  if (detail::has_nested_columns(t)) {
-    auto [dremel_data, d_dremel_device_view] = list_lex_preprocess(verticalized_lhs, stream);
+  if (detail::has_nested_columns(preprocessed_input)) {
+    auto [dremel_data, d_dremel_device_view] = list_lex_preprocess(preprocessed_input, stream);
     return std::shared_ptr<preprocessed_table>(
-      new preprocessed_table(std::move(d_t),
+      new preprocessed_table(std::move(d_table),
                              std::move(d_column_order),
                              std::move(d_null_precedence),
                              std::move(d_depths),
                              std::move(dremel_data),
-                             std::move(d_dremel_device_view)));
+                             std::move(d_dremel_device_view),
+                             std::move(transformed_columns),
+                             ranked_children));
   } else {
-    return std::shared_ptr<preprocessed_table>(new preprocessed_table(std::move(d_t),
-                                                                      std::move(d_column_order),
-                                                                      std::move(d_null_precedence),
-                                                                      std::move(d_depths)));
+    return std::shared_ptr<preprocessed_table>(
+      new preprocessed_table(std::move(d_table),
+                             std::move(d_column_order),
+                             std::move(d_null_precedence),
+                             std::move(d_depths),
+                             std::move(transformed_columns),
+                             ranked_children));
   }
 }
 
+std::shared_ptr<preprocessed_table> preprocessed_table::create(
+  table_view const& input,
+  host_span<order const> column_order,
+  host_span<null_order const> null_precedence,
+  rmm::cuda_stream_view stream)
+{
+  auto [decomposed_input, new_column_order, new_null_precedence, verticalized_col_depths] =
+    decompose_structs(input, false /*no decompose lists*/, column_order, null_precedence);
+
+  // Unused variables are generated for rhs table which is not available here.
+  [[maybe_unused]] auto [transformed_input, unused_0, transformed_columns, unused_1] =
+    transform_lists_of_structs(decomposed_input, std::nullopt, new_null_precedence, stream);
+
+  auto const ranked_children = transformed_columns.size() > 0;
```

Comment:
I prefer the name `has_ranked_children` to indicate it's a boolean and not something that manages columns.
```suggestion
  auto const has_ranked_children = !transformed_columns.empty();
```